### PR TITLE
soundwire: intel: transition to 3 steps initialization

### PR DESCRIPTION
**Pull request by bardliao**
The PR has the same commits as https://github.com/thesofproject/linux/pull/2107 and target to a new base which aligned with upstream. 
@plbossart All the commends in https://github.com/thesofproject/linux/pull/2107/commits/5d1b316a59a0943891a5edc571a7de4b8e0c0751 are applied except the setting sdw->cdns.registers in sdw_intel_probe_controller() one. I am going to send it to upstream directly and hope it can be reviewed/merged soon.

### CODE REVIEW

File: drivers/soundwire/intel_init.c [hunk 1]
```diff
@@ -23,22 +24,51 @@
...
 	int i;
 
 	if (!link)
 		return 0;
 
-	for (i = 0; i < ctx->count; i++) {
+	link_mask = ctx->link_mask;
+
+	for (i = 0; i < ctx->count; i++, link++) {
+		if (link_mask && !(link_mask & BIT(i)))
```
Comment (RanderWang) [inline drivers/soundwire/intel_init.c:67]: link_mask is defined out of loop and unchanged, so it is better to move it out, such as
````
if (!link_mask)
    goto kfree
````
then you can modify it to 
````
if (link_mask & BIT (i) && link->dev)
    platform_device_unregister(link->pdev);
````

Comment (RanderWang) [reply drivers/soundwire/intel_init.c:67]: tune it again.   we can just do it as 
````
if (!(link_mask & BIT(i)))
    continue;
````
if link_mask = 0,  !(link_mask & BIT(i)) = 1, so continue
And we don't need to add "int link_mask"

File: drivers/soundwire/intel_init.c [hunk 2]
```diff
@@ -47,52 +77,97 @@ static int sdw_intel_cleanup_pdev(struct sdw_intel_ctx *ctx)
...
-		return NULL;
+		return -EINVAL;
+	}
+	dev_dbg(&adev->dev, "ACPI reports %d SDW Link devices\n", count);
+
+	info->count = count;
+	info->link_mask = 0;
+
+	for (i = 0; i < count; i++) {
+		if (ctrl_link_mask && !(ctrl_link_mask & BIT(i))) {
```
Comment (RanderWang) [inline drivers/soundwire/intel_init.c:130]: And also here, remove ctrl_link_mask check as above comment

File: drivers/soundwire/intel_init.c [hunk 3]
```diff
@@ -47,52 +77,97 @@ static int sdw_intel_cleanup_pdev(struct sdw_intel_ctx *ctx)
...
+			dev_dbg(&adev->dev,
+				"Link %d masked, will not be enabled\n", i);
+			continue;
+		}
+
+		if (!is_link_enabled(acpi_fwnode_handle(adev), i)) {
+			dev_dbg(&adev->dev,
+				"Link %d not selected in firmware\n", i);
```
Comment (RanderWang) [inline drivers/soundwire/intel_init.c:138]: I don't know whether it should be "Link %d is not selected in firmware\n" or not. I am a beginner in English.

File: drivers/soundwire/intel_init.c [hunk 4]
```diff
@@ -104,19 +179,22 @@ static struct sdw_intel_ctx
...
+	link_mask = ctx->link_mask;
 
 	/* Create SDW Master devices */
-	for (i = 0; i < count; i++) {
-		if (link_mask && !(link_mask & BIT(i))) {
-			dev_dbg(&adev->dev,
-				"Link %d masked, will not be enabled\n", i);
-			link++;
+	for (i = 0; i < count; i++, link++) {
+		if (link_mask && !(link_mask & BIT(i)))
```
Comment (RanderWang) [inline drivers/soundwire/intel_init.c:192]: also remove link_mask check

File: drivers/soundwire/intel_init.c [hunk 5]
```diff
@@ -135,26 +213,68 @@ static struct sdw_intel_ctx
...
 		link->pdev = pdev;
-		link++;
 	}
 
 	return ctx;
 
-pdev_err:
-	sdw_intel_cleanup_pdev(ctx);
+err:
+	ctx->count = i;
```
Comment (RanderWang) [inline drivers/soundwire/intel_init.c:227]: if err,  ctx->count is used for ?

File: drivers/soundwire/intel_init.c [hunk 6]
```diff
@@ -135,26 +213,68 @@ static struct sdw_intel_ctx
...
+	}
+
+	if (!ctx->links)
+		return -EINVAL;
+
+	link = ctx->links;
+	link_mask = ctx->link_mask;
+
+	/* Startup SDW Master devices */
+	for (i = 0; i < ctx->count; i++, link++) {
```
Comment (RanderWang) [inline drivers/soundwire/intel_init.c:264]: we don't need link_mask and its check for NULL

**Comment by bardliao** [comment]
@RanderWang Sorry, but the code is from current topic/sof-dev branch. You may want to submit a PR to fix them :)

**Comment by bardliao** [comment]
@plbossart @ranj063 I addressed Vinod's commands. Could you help to review? https://www.mail-archive.com/linux-kernel@vger.kernel.org/msg2167056.html

**Comment by plbossart** [comment]
@bardliao please look at PR #2134 and tag sof-dev-rebase-20200521, I think all your changes are already added. If not let's add them.

**Pull request closed without merging**